packages/resolve-url-loader-filesearch/index.js.js

```diff
@@ -4,151 +4,78 @@
-
-    // the queue pattern ensures that we favour paths closest the the start path
-    // process the queue until empty or until we exhaust our attempts
-    while (queue.length && (remaining-- > 0)) {
-
-      // shift the first item off the queue, consider it the base for our relative uri
-      var basePath = queue.shift();
-      var fullPath = path.resolve(basePath, uri);
-      messages.push(basePath);
-
-      // file exists so end
-      if (fs.existsSync(fullPath)) {
-        flushMessages('FOUND');
-        return basePath;
-      }
-      // enqueue subdirectories that are not packages and are not in the root path
-      else {
-        enqueue(queue, basePath);
-      }
-    }
-
-    // interrupted by options.attempts
-    if (queue.length) {
-      flushMessages('NOT FOUND (INTERRUPTED)');
-    }
-    // not found
-    else {
-      flushMessages('NOT FOUND');
-      return null;
-    }
-
-    /**
-     * Enqueue subdirectories that are not packages and are not in the root path
-     * @param {Array} queue The queue to add to
-     * @param {string} basePath The path to consider
-     */
-    function enqueue(queue, basePath) {
-      fs.readdirSync(basePath)
-        .filter(function notHidden(filename) {
-          return (filename.charAt(0) !== '.');
-        })
-        .map(function toAbsolute(filename) {
-          return path.join(basePath, filename);
-        })
-        .filter(function directoriesOnly(absolutePath) {
-          return fs.existsSync(absolutePath) && fs.statSync(absolutePath).isDirectory();
-        })
-        .filter(function notInRootPath(absolutePath) {
-          return (pathToRoot.indexOf(absolutePath) < 0);
-        })
-        .filter(testNotPackage)
-        .forEach(function enqueue(absolutePath) {
-          queue.push(absolutePath);
-        });
-    }
-
-    /**
-     * Test whether the given directory is above but not equal to any of the project root directories.
-     * @param {string} absolutePath An absolute path
-     * @returns {boolean} True where a package.json or bower.json exists, else False
-     */
-    function testWithinLimit(absolutePath) {
+const path = require('path');
+const { createJoinFunction, defaultJoinGenerator, defaultJoinOperation } = require('resolve-url-loader');
+
+module.exports = (options = {}) => {
+  const includeRoot = !!options.includeRoot;
+  const attempts = Math.max(0, options.attempts) || 1E+9;
+  const breakOnFile = [].concat(options.breakOnFile) || ['package.json', 'bower.json'];
+
+  const predicate = typeof options.predicate === 'function' ?
+    options.testIsRunning :
+    ((fs, absolutePath) => breakOnFile
+      .map((file) => path.resolve(absolutePath, file))
+      .every((file) => !fs.existsSync(file) || !fs.statSync(file).isFile()));
+
+  const baseGenerator = typeof options.generator === 'function' ?
+    options.generator :
+    defaultJoinGenerator;
+
+  const searchingGeneartor = (filename, uri, bases, isAbsolute, {root, fs}) => {
```

Comment:
@runfaj @mightypork this is untested but it is the original file-search "magic" refactored as a generator.

I'm open to ideas on what we can do (with minimum effort) to make this available/useful. And if we do what level of testing is appropriate.